bin/cron/build_contact_rollups

```diff
@@ -0,0 +1,428 @@
+            WHERE course IN (#{COURSE_LIST})
+            UNION
+            SELECT teacher_id,
+              CASE program_type
+                WHEN 1 THEN 'CS in Science'
+                WHEN 2 THEN 'CS in Algebra'
+                WHEN 3 THEN 'Exploring Computer Science'
+                WHEN 4 THEN 'CS Principles'
+                WHEN 5 THEN 'CS Fundamentals'
+              END AS course
+            FROM #{DASHBOARD_DB_NAME}.workshop_attendance AS workshop_attendance
+              INNER JOIN #{DASHBOARD_DB_NAME}.segments AS segments ON segments.id = workshop_attendance.segment_id
+              INNER JOIN #{DASHBOARD_DB_NAME}.workshops AS workshops ON workshops.id = segments.workshop_id
+            ) q1
+          ORDER BY 1,2
+          ) q2
+        GROUP BY teacher_id
+      ) src
+    SET contact_rollups.professional_learning_attended = src.courses
+    WHERE contact_rollups.dashboard_user_id = src.teacher_id"
+  log "   Completed in " + ((Time.now) - start).to_s + " seconds"
+end
+
+def mysql_multi_connection
+  # return a connection with the MULTI_STATEMENTS flag set that allows multiple statements in one DB call
+  Sequel.connect(CDO.pegasus_db_writer.sub('mysql:', 'mysql2:'), flags: ::Mysql2::Client::MULTI_STATEMENTS)
+end
+
+def update_geo_data_from_forms(form_kind)
+  start = Time.now
+  log "Updating geo data for form kind #{form_kind}"
+
+  record_count = 0
+  num_updated = 0
+  num_parse_errors = 0
+  update_batch = ""
+  conn = mysql_multi_connection
+  time_last_output = start
+
+  PEGASUS_DB[:forms].where(kind: form_kind).each do |form|
+    record_count += 1
+    begin
+      data = JSON.parse(form[:data])
+    rescue JSON::ParserError
+      # we have just a couple of records with invalid JSON (duplicate fields), rescue and skip these
+      num_parse_errors += 1
+      next
+    end
+
+    email = data['email_s'].presence
+    next if email.nil?
+
+    # get user-supplied address/location data from form if present, from any of the differently-named location fields across all form kinds
+    street_address = data['location_street_address_s'].presence || data['user_street_address_s'].presence
+    city = data['location_city_s'].presence || data['user_city_s'].presence || data['teacher_city_s'].presence
+    state = data['location_state_s'].presence || data['teacher_state_s'].presence
+    if state.nil?
+      # note that the 'user_state_s' record is in fact a *state code*, not a state name
+      state_code = data['location_state_code_s'].presence || data['state_code_s'].presence || data['user_state_s'].presence
+      state = get_us_state_from_abbr(abbr: state_code, include_dc: true).presence unless state_code.nil?
+    end
+    postal_code = data['location_postal_code_s'].presence || data['zip_code_s'].presence || data['user_postal_code_s'].presence
+    country = data['location_country_s'].presence || data['country_s'].presence || data['teacher_country_s'].presence
```

Comment:
I thought about if I could make this any cleaner. One catch is the field names are specific to the forms and don't quite follow a pattern (for instance, one field saved as 'state' is really state code). I could make a helper that given an array of field names would return the first one with presence. That makes 6 lines of code a little shorter while adding 3-6 lines for the helper function, so net increase in lines of code. Didn't seem worth it. If you have any clever ideas I'm all ears!